File: android/src/main/java/expo/modules/aikit/GemmaInferenceClient.kt
package expo.modules.aikit

import android.app.ActivityManager
import android.content.Context
import android.os.Build
import com.google.ai.edge.litertlm.Engine
import com.google.ai.edge.litertlm.EngineConfig
import com.google.ai.edge.litertlm.Conversation
import com.google.ai.edge.litertlm.ConversationConfig
import com.google.ai.edge.litertlm.Backend
import com.google.ai.edge.litertlm.Message
import com.google.ai.edge.litertlm.MessageCallback
import com.google.ai.edge.litertlm.Contents
import com.google.ai.edge.litertlm.SamplerConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.coroutines.suspendCancellableCoroutine
import java.io.File
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * Wrapper around LiteRT-LM Engine for Gemma 4 models.
 *
 * Concurrency model:
 * - A Mutex guards all state transitions (load, unload, inference).
 * - sendMessage/startStreaming block on the mutex if a load is in progress.
 * - deleteModel waits for inference to finish, then unloads and deletes.
 * - A separate isDownloading flag prevents concurrent downloads (checked before
 *   the long-running download, not inside the mutex).
 */
class GemmaInferenceClient(private val context: Context) {

  private val mutex = Mutex()
  private var engine: Engine? = null
  private var conversation: Conversation? = null
  private var loadedModelId: String? = null

  // Sampling config for the session, fixed at load. Kept so each inference call
  // can build a FRESH conversation: the JS layer passes the full history every
  // call (stateless-model contract), so reusing one native conversation would
  // feed the model its own history twice, and a second turn on the same
  // conversation trips LiteRT-LM 0.10.0's chat-template engine on some models
  // (qwen3: "string has no method named strip").
  private var convConfig: ConversationConfig = ConversationConfig()

  @Volatile
  private var isDownloading = false

  @Volatile
  private var cancelDownloadRequested = false

  // -------------------------------------------------------------------------
  // Model lifecycle
  // -------------------------------------------------------------------------

  /**
   * Fail fast on ABIs where LiteRT-LM's native backend is known to crash.
   *
   * litertlm-android ships an x86_64 liblitertlm_jni.so, but the x86_64 backend
   * is unvalidated upstream and SIGSEGVs on emulators, at engine init and
   * during inference (google-ai-edge/LiteRT-LM #2799, #2159; still open as of
   * 0.15.0). A SIGSEGV kills the whole app process and no Kotlin try/catch can
   * contain it, so the only safe behavior is a typed error before any Engine
   * call. Physical devices and arm64 emulator images (Apple Silicon hosts) run
   * the arm64-v8a library and are unaffected. Remove this guard only once a
   * LiteRT-LM release validates x86_64.
   */
  private fun requireSupportedAbi(modelId: String) {
    val primaryAbi = Build.SUPPORTED_ABIS.firstOrNull() ?: return
    if (primaryAbi == "x86_64" || primaryAbi == "x86") {
      throw RuntimeException(
        "DEVICE_NOT_SUPPORTED:$modelId:Downloadable models are not supported on $primaryAbi Android " +
        "(typically an emulator on an Intel/AMD host), LiteRT-LM's native x86 backend crashes the app. " +
        "Use a physical device or an arm64 emulator image."
      )
    }
  }

  /**
   * Load a model into memory using LiteRT-LM Engine.
   * Unloads any previously loaded model first.
   * Caller is responsible for emitting onModelStateChange events.
   */
  suspend fun loadModel(
    modelId: String,
    modelPath: String,
    minRamBytes: Long = 0,
    backend: String = "auto",
    temperature: Double? = null,
    topK: Int? = null,
    topP: Double? = null
  ) = mutex.withLock {
    requireSupportedAbi(modelId)

    // Unload previous model if different
    if (loadedModelId != null && loadedModelId != modelId) {
      conversation?.close()
      engine?.close()
      conversation = null
      engine = null
      loadedModelId = null
    }

    if (loadedModelId == modelId && engine != null) {
      return@withLock // Already loaded
    }

    // Soft memory check. LiteRT-LM memory-maps model weights so actual RSS
    // is much lower than file size. We log a warning but always attempt the
    // load, Engine.initialize() may still succeed. If it truly OOMs, Android's
    // lmkd kills the process (uncatchable signal 9), but that's better than
    // blocking devices that could have worked.
    val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager
    if (activityManager != null && minRamBytes > 0) {
      val memInfo = ActivityManager.MemoryInfo()
      activityManager.getMemoryInfo(memInfo)
      if (memInfo.availMem < minRamBytes) {
        android.util.Log.w("ExpoAiKit",
          "Low memory warning for $modelId: " +
          "available ${memInfo.availMem / 1_000_000}MB, recommended ${minRamBytes / 1_000_000}MB. " +
          "Attempting load anyway (LiteRT-LM uses memory-mapped I/O)."
        )
      }
    }

    try {
      withContext(Dispatchers.IO) {
        val newEngine = when (backend) {
          "gpu" -> {
            val config = EngineConfig(modelPath = modelPath, backend = Backend.GPU())
            val eng = Engine(config)
            eng.initialize()
            eng
          }
          "cpu" -> {
            val config = EngineConfig(modelPath = modelPath, backend = Backend.CPU())
            val eng = Engine(config)
            eng.initialize()
            eng
          }
          else -> {
            // "auto": try GPU first, fall back to CPU
            try {
              val gpuConfig = EngineConfig(modelPath = modelPath, backend = Backend.GPU())
              val eng = Engine(gpuConfig)
              eng.initialize()
              eng
            } catch (e: Exception) {
              val cpuConfig = EngineConfig(modelPath = modelPath, backend = Backend.CPU())
              val eng = Engine(cpuConfig)
              eng.initialize()
              eng
            }
          }
        }

        // Sampling knobs are fixed at conversation creation by LiteRT-LM. If any
        // is provided, build a SamplerConfig (filling unspecified values with
        // Gemma-typical defaults); otherwise use the engine/model defaults.
        // Stored so every inference call can create a fresh conversation with
        // the same sampling (see convConfig).
        convConfig = if (temperature != null || topK != null || topP != null) {
          ConversationConfig(
            samplerConfig = SamplerConfig(
              topK = topK ?: 64,
              topP = topP ?: 0.95,
              temperature = temperature ?: 1.0
            )
          )
        } else {
          ConversationConfig()
        }
        val newConversation = newEngine.createConversation(convConfig)

        engine = newEngine
        conversation = newConversation
        loadedModelId = modelId
      }
    } catch (e: OutOfMemoryError) {
      conversation?.close()
      engine?.close()
      conversation = null
      engine = null
      loadedModelId = null
      throw RuntimeException("INFERENCE_OOM:$modelId:Device does not have enough memory to load model")
    } catch (e: Exception) {
      conversation?.close()
      engine?.close()
      conversation = null
      engine = null
      loadedModelId = null
      throw RuntimeException("MODEL_LOAD_FAILED:$modelId:${e.message}")
    }
  }

  /**
   * Unload the current model from memory.
   */
  suspend fun unloadModel() = mutex.withLock {
    conversation?.close()
    engine?.close()
    conversation = null
    engine = null
    loadedModelId = null
  }

  fun getLoadedModelId(): String? = loadedModelId

  fun isModelLoaded(): Boolean = engine != null

  // -------------------------------------------------------------------------
  // Inference
  // -------------------------------------------------------------------------

  /**
   * Replace the current conversation with a fresh one for a single call.
   * Must be called with the mutex held. See [convConfig] for why inference
   * never reuses a conversation across calls.
   */
  private fun freshConversation(): Conversation {
    val eng = engine
      ?: throw RuntimeException("MODEL_NOT_DOWNLOADED:${loadedModelId ?: "unknown"}:No model loaded")
    conversation?.close()
    conversation = null
    val conv = try {
      eng.createConversation(convConfig)
    } catch (e: Exception) {
      throw RuntimeException("INFERENCE_FAILED:${loadedModelId ?: "unknown"}:Failed to create conversation: ${e.message}")
    }
    conversation = conv
    return conv
  }

  /**
   * Generate a complete response. Blocks until done.
   * The mutex ensures this cannot run concurrently with load/unload.
   */
  suspend fun generateText(prompt: String, systemPrompt: String): String = mutex.withLock {
    val conv = freshConversation()

    val fullPrompt = buildFullPrompt(prompt, systemPrompt)

    try {
      withContext(Dispatchers.IO) {
        suspendCancellableCoroutine<String> { continuation ->
          val result = StringBuilder()
          var previousText = ""
          conv.sendMessageAsync(
            Contents.of(fullPrompt),
            object : MessageCallback {
              override fun onMessage(message: Message) {
                // LiteRT-LM may deliver accumulated text or delta tokens
                // (observed: deltas on-device), detect which, exactly like
                // generateTextStream, so the final text isn't just the last chunk.
                val messageText = message.toString()
                if (messageText.startsWith(previousText) && messageText.length >= previousText.length) {
                  result.clear()
                  result.append(messageText)
                } else {
                  result.append(messageText)
                }
                previousText = result.toString()
              }
              override fun onDone() {
                continuation.resume(result.toString())
              }
              override fun onError(throwable: Throwable) {
                continuation.resumeWithException(throwable)
              }
            },
            emptyMap()
          )
        }
      }
    } catch (e: kotlinx.coroutines.CancellationException) {
      // Cooperative cancellation, propagate, don't mask as an inference failure.
      throw e
    } catch (e: OutOfMemoryError) {
      throw RuntimeException("INFERENCE_OOM:${loadedModelId ?: "unknown"}:Out of memory during inference")
    } catch (e: Exception) {
      throw RuntimeException("INFERENCE_FAILED:${loadedModelId ?: "unknown"}:${e.message}")
    }
  }

  /**
   * Generate a streaming response. The onChunk callback receives
   * (token=delta, accumulatedText=full, isDone) matching the PromptApiClient contract.
   *
   * LiteRT-LM 0.10.0 uses a MessageCallback interface. Each onMessage emission
   * contains accumulated text, so we diff against previousText to extract
   * the delta token.
   */
  suspend fun generateTextStream(
    prompt: String,
    systemPrompt: String,
    onChunk: (token: String, accumulatedText: String, isDone: Boolean) -> Unit
  ) = mutex.withLock {
    val conv = freshConversation()

    val fullPrompt = buildFullPrompt(prompt, systemPrompt)

    try {
      withContext(Dispatchers.IO) {
        suspendCancellableCoroutine<Unit> { continuation ->
          val accumulatedBuilder = StringBuilder()
          var previousText = ""
          conv.sendMessageAsync(
            Contents.of(fullPrompt),
            object : MessageCallback {
              override fun onMessage(message: Message) {
                val messageText = message.toString()

                // LiteRT-LM may deliver accumulated text or delta tokens depending
                // on the version. Detect which by checking if messageText extends
                // what we've seen before.
                val token: String
                if (messageText.startsWith(previousText) && messageText.length >= previousText.length) {
                  // Accumulated text, extract delta
                  token = messageText.substring(previousText.length)
                  previousText = messageText
                  accumulatedBuilder.clear()
                  accumulatedBuilder.append(messageText)
                } else {
                  // Delta token, accumulate ourselves
                  token = messageText
                  accumulatedBuilder.append(messageText)
                  previousText = accumulatedBuilder.toString()
                }

                val accumulated = accumulatedBuilder.toString()
                onChunk(token, accumulated, false)
              }
              override fun onDone() {
                val finalText = accumulatedBuilder.toString()
                onChunk("", finalText, true)
                continuation.resume(Unit)
              }
              override fun onError(throwable: Throwable) {
                continuation.resumeWithException(throwable)
              }
            },
            emptyMap()
          )
        }
      }
    } catch (e: kotlinx.coroutines.CancellationException) {
      // Cooperative cancellation, propagate, don't mask as an inference failure.
      throw e
    } catch (e: OutOfMemoryError) {
      throw RuntimeException("INFERENCE_OOM:${loadedModelId ?: "unknown"}:Out of memory during inference")
    } catch (e: Exception) {
      throw RuntimeException("INFERENCE_FAILED:${loadedModelId ?: "unknown"}:${e.message}")
    }
  }

  // -------------------------------------------------------------------------
  // Download
  // -------------------------------------------------------------------------

  /**
   * Download a model file with progress reporting.
   * Prevents concurrent downloads. On failure, deletes partial files.
   *
   * Strategy: restart from scratch on failure (no HTTP Range resumption).
   * Downloads to a .tmp file, verifies SHA-256, atomically renames on success
   * (the mechanics live in [DownloadUtil], shared with the embedding asset).
   */
  suspend fun downloadModelFile(
    modelId: String,
    url: String,
    sha256: String,
    onProgress: (bytesRead: Long, totalBytes: Long) -> Unit
  ) {
    if (isDownloading) {
      throw RuntimeException("DOWNLOAD_FAILED:$modelId:Download already in progress")
    }
    isDownloading = true
    cancelDownloadRequested = false

    try {
      withContext(Dispatchers.IO) {
        val modelsDir = File(context.filesDir, "models")
        modelsDir.mkdirs()

        DownloadUtil.downloadVerified(
          context = context,
          modelId = modelId,
          url = url,
          targetFile = File(modelsDir, "$modelId.litertlm"),
          tempFile = File(modelsDir, "$modelId.litertlm.tmp"),
          sha256 = sha256,
          isCancelled = { cancelDownloadRequested },
          onProgress = onProgress
        )
      }
    } finally {
      isDownloading = false
    }
  }

  /**
   * Delete a model file from disk. If the model is loaded, unloads it first.
   */
  suspend fun deleteModelFile(modelId: String) = mutex.withLock {
    // Unload if this model is currently loaded
    if (loadedModelId == modelId) {
      conversation?.close()
      engine?.close()
      conversation = null
      engine = null
      loadedModelId = null
    }

    val modelFile = File(context.filesDir, "models/$modelId.litertlm")
    if (modelFile.exists()) {
      modelFile.delete()
    }
    // Also clean up any partial downloads
    val tempFile = File(context.filesDir, "models/$modelId.litertlm.tmp")
    if (tempFile.exists()) {
      tempFile.delete()
    }
  }

  /**
   * Request cancellation of the in-flight download, if any. The download loop
   * checks this flag and throws DOWNLOAD_CANCELLED. No-op if nothing is downloading.
   */
  fun cancelDownload(modelId: String) {
    cancelDownloadRequested = true
  }

  /**
   * Check if a model file exists on disk.
   */
  fun isModelFileDownloaded(modelId: String): Boolean {
    return File(context.filesDir, "models/$modelId.litertlm").exists()
  }

  /**
   * Get the file path for a downloaded model.
   */
  fun getModelFilePath(modelId: String): String {
    return File(context.filesDir, "models/$modelId.litertlm").absolutePath
  }

  // -------------------------------------------------------------------------
  // Private helpers
  // -------------------------------------------------------------------------

  private fun buildFullPrompt(prompt: String, systemPrompt: String): String {
    return if (systemPrompt.isNotBlank()) {
      "$systemPrompt\n\n$prompt"
    } else {
      prompt
    }
  }
}

File: ios/GemmaInferenceClient.swift
import Foundation
import CryptoKit

actor GemmaInferenceClient {

  private var engine: Engine?
  private var conversation: Conversation?
  private var loadedModelId: String?

  private var isDownloading = false
  private var currentDownloader: ModelDownloader?

  // MARK: - Paths

  private static var modelsDirectory: URL {
    let support = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
    let dir = support.appendingPathComponent("ExpoAiKit/Models", isDirectory: true)
    try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    return dir
  }

  private static var cacheDirectory: URL {
    let caches = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
    let dir = caches.appendingPathComponent("ExpoAiKit/litertlm", isDirectory: true)
    try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    return dir
  }

  nonisolated func modelFileURL(_ modelId: String) -> URL {
    return Self.modelsDirectory.appendingPathComponent("\(modelId).litertlm")
  }

  nonisolated func isModelFileDownloaded(_ modelId: String) -> Bool {
    return FileManager.default.fileExists(atPath: modelFileURL(modelId).path)
  }

  func getLoadedModelId() -> String? { loadedModelId }
  func isModelLoaded() -> Bool { engine != nil }

  // MARK: - Model lifecycle

  /// Load a model into memory. Unloads any previously loaded model first.
  ///
  /// Sampling knobs (temperature/topK/topP/seed) are fixed at conversation
  /// creation by LiteRT-LM, so they're applied here rather than per-call. If any
  /// is provided, a full SamplerConfig is built (LiteRT-LM requires all three of
  /// topK/topP/temperature), filling unspecified values with Gemma-typical
  /// defaults; if none is provided, the engine/model defaults are used.
  func loadModel(
    modelId: String,
    backend: String,
    temperature: Float? = nil,
    topK: Int? = nil,
    topP: Float? = nil,
    seed: Int? = nil
  ) async throws {
    // Unload previous model if different
    if let current = loadedModelId, current != modelId {
      conversation = nil
      engine = nil
      loadedModelId = nil
    }
    if loadedModelId == modelId, engine != nil {
      return // Already loaded
    }

    let modelPath = modelFileURL(modelId).path
    guard FileManager.default.fileExists(atPath: modelPath) else {
      throw GemmaError.modelNotDownloaded(modelId)
    }

    let cacheDir = Self.cacheDirectory.path

    do {
      let newEngine = try await initializeEngine(
        modelPath: modelPath, cacheDir: cacheDir, backend: backend
      )
      let samplerConfig: SamplerConfig?
      if temperature != nil || topK != nil || topP != nil || seed != nil {
        samplerConfig = try SamplerConfig(
          topK: topK ?? 64,
          topP: topP ?? 0.95,
          temperature: temperature ?? 1.0,
          seed: seed ?? 0
        )
      } else {
        samplerConfig = nil
      }
      let newConversation = try await newEngine.createConversation(
        with: ConversationConfig(samplerConfig: samplerConfig)
      )
      engine = newEngine
      conversation = newConversation
      loadedModelId = modelId
    } catch {
      conversation = nil
      engine = nil
      loadedModelId = nil
      throw GemmaError.modelLoadFailed(modelId, reason: "\(error)")
    }
  }

  /// "auto" tries GPU first, then falls back to CPU. "gpu"/"cpu" force the backend.
  private func initializeEngine(modelPath: String, cacheDir: String, backend: String) async throws -> Engine {
    func makeEngine(_ b: Backend) async throws -> Engine {
      let config = try EngineConfig(modelPath: modelPath, backend: b, cacheDir: cacheDir)
      let engine = Engine(engineConfig: config)
      try await engine.initialize()
      return engine
    }

    switch backend {
    case "gpu":
      return try await makeEngine(.gpu)
    case "cpu":
      return try await makeEngine(.cpu())
    default:
      do { return try await makeEngine(.gpu) }
      catch { return try await makeEngine(.cpu()) }
    }
  }

  func unloadModel() async {
    conversation = nil
    engine = nil
    loadedModelId = nil
  }

  // MARK: - Inference

  /// Generate a complete response. Blocks until done.
  func generateText(prompt: String, systemPrompt: String) async throws -> String {
    guard let conv = conversation else {
      throw GemmaError.modelNotDownloaded(loadedModelId ?? "unknown")
    }
    let fullPrompt = buildFullPrompt(prompt: prompt, systemPrompt: systemPrompt)
    do {
      let response = try await conv.sendMessage(Message(fullPrompt))
      return response.toString
    } catch {
      throw GemmaError.inferenceFailed(loadedModelId ?? "unknown", reason: "\(error)")
    }
  }

  /// Generate a streaming response. The onChunk closure receives
  /// (token=delta, accumulatedText=full, isDone) matching the contract used by
  /// startStreaming in ExpoAiKitModule.swift.
  ///
  /// LiteRT-LM may deliver each Message chunk as either an accumulated string
  /// or a delta — mirror Android's detection logic to handle both safely.
  func generateTextStream(
    prompt: String,
    systemPrompt: String,
    onChunk: @Sendable (_ token: String, _ accumulatedText: String, _ isDone: Bool) -> Void
  ) async throws {
    guard let conv = conversation else {
      throw GemmaError.modelNotDownloaded(loadedModelId ?? "unknown")
    }
    let fullPrompt = buildFullPrompt(prompt: prompt, systemPrompt: systemPrompt)

    var accumulated = ""
    var previousText = ""

    do {
      let stream = conv.sendMessageStream(Message(fullPrompt))
      for try await chunk in stream {
        if Task.isCancelled {
          try? conv.cancel()
          break
        }

        let chunkText = chunk.toString
        let token: String
        if chunkText.hasPrefix(previousText) && chunkText.count >= previousText.count {
          // Accumulated text — extract delta
          token = String(chunkText.dropFirst(previousText.count))
          previousText = chunkText
          accumulated = chunkText
        } else {
          // Delta token — accumulate ourselves
          token = chunkText
          accumulated += chunkText
          previousText = accumulated
        }
        onChunk(token, accumulated, false)
      }
      // Always emit a terminal chunk — including on cooperative cancellation —
      // so the JS stream settles instead of hanging.
      onChunk("", accumulated, true)
    } catch is CancellationError {
      try? conv.cancel()
      onChunk("", accumulated, true)
    } catch {
      throw GemmaError.inferenceFailed(loadedModelId ?? "unknown", reason: "\(error)")
    }
  }

  private nonisolated func buildFullPrompt(prompt: String, systemPrompt: String) -> String {
    let trimmed = systemPrompt.trimmingCharacters(in: .whitespacesAndNewlines)
    return trimmed.isEmpty ? prompt : "\(trimmed)\n\n\(prompt)"
  }

  // MARK: - Download

  /// Download a model file with progress reporting. Atomic via .tmp → rename.
  func downloadModelFile(
    modelId: String,
    urlString: String,
    sha256: String,
    onProgress: @Sendable @escaping (_ bytesRead: Int64, _ totalBytes: Int64) -> Void
  ) async throws {
    if isDownloading {
      throw GemmaError.downloadFailed(modelId, reason: "Download already in progress")
    }
    isDownloading = true
    defer {
      isDownloading = false
      currentDownloader = nil
    }

    guard let url = URL(string: urlString) else {
      throw GemmaError.downloadFailed(modelId, reason: "Invalid URL")
    }

    let targetURL = modelFileURL(modelId)
    let tempURL = targetURL.appendingPathExtension("tmp")

    // Clean any prior partial
    try? FileManager.default.removeItem(at: tempURL)

    do {
      let downloader = ModelDownloader(progressHandler: onProgress)
      currentDownloader = downloader
      let downloadedURL = try await downloader.download(from: url)

      // Move downloaded file to .tmp location
      try? FileManager.default.removeItem(at: tempURL)
      try FileManager.default.moveItem(at: downloadedURL, to: tempURL)

      // SHA256 verify
      if !sha256.isEmpty {
        let actual = try sha256OfFile(at: tempURL)
        if actual.lowercased() != sha256.lowercased() {
          try? FileManager.default.removeItem(at: tempURL)
          throw GemmaError.downloadCorrupt(modelId, expected: sha256, actual: actual)
        }
      }

      // Atomic rename
      try? FileManager.default.removeItem(at: targetURL)
      try FileManager.default.moveItem(at: tempURL, to: targetURL)
    } catch let error as GemmaError {
      try? FileManager.default.removeItem(at: tempURL)
      throw error
    } catch let urlError as URLError where urlError.code == .cancelled {
      try? FileManager.default.removeItem(at: tempURL)
      throw GemmaError.downloadCancelled(modelId)
    } catch {
      try? FileManager.default.removeItem(at: tempURL)
      throw GemmaError.downloadFailed(modelId, reason: "\(error)")
    }
  }

  /// Cancel the in-flight download, if any. The pending downloadModelFile call
  /// throws GemmaError.downloadCancelled.
  func cancelDownload(_ modelId: String) {
    currentDownloader?.cancel()
  }

  /// Delete a model file. If the model is currently loaded, unloads it first.
  func deleteModelFile(modelId: String) async {
    if loadedModelId == modelId {
      conversation = nil
      engine = nil
      loadedModelId = nil
    }
    let target = modelFileURL(modelId)
    let temp = target.appendingPathExtension("tmp")
    try? FileManager.default.removeItem(at: target)
    try? FileManager.default.removeItem(at: temp)
  }

  // MARK: - Helpers

  private nonisolated func sha256OfFile(at url: URL) throws -> String {
    let handle = try FileHandle(forReadingFrom: url)
    defer { try? handle.close() }
    var hasher = SHA256()
    while true {
      guard let chunk = try handle.read(upToCount: 1 << 20), !chunk.isEmpty else { break }
      autoreleasepool { hasher.update(data: chunk) }
    }
    let digest = hasher.finalize()
    return digest.map { String(format: "%02x", $0) }.joined()
  }
}

// MARK: - URL session download wrapper

private final class ModelDownloader: NSObject, URLSessionDownloadDelegate, @unchecked Sendable {
  private var continuation: CheckedContinuation<URL, Error>?
  private let progressHandler: @Sendable (Int64, Int64) -> Void
  private var didFinish = false
  private var task: URLSessionDownloadTask?

  init(progressHandler: @Sendable @escaping (Int64, Int64) -> Void) {
    self.progressHandler = progressHandler
  }

  func download(from url: URL) async throws -> URL {
    return try await withCheckedThrowingContinuation { (cont: CheckedContinuation<URL, Error>) in
      self.continuation = cont
      let session = URLSession(configuration: .default, delegate: self, delegateQueue: nil)
      var request = URLRequest(url: url)
      request.timeoutInterval = 60
      let task = session.downloadTask(with: request)
      self.task = task
      task.resume()
    }
  }

  /// Cancel the download. The delegate's didCompleteWithError fires with
  /// URLError.cancelled, which downloadModelFile maps to downloadCancelled.
  func cancel() {
    task?.cancel()
  }

  func urlSession(
    _ session: URLSession,
    downloadTask: URLSessionDownloadTask,
    didFinishDownloadingTo location: URL
  ) {
    // Move to a stable temp location synchronously, before this delegate method returns
    // (the system deletes `location` after we return).
    let stableTemp = FileManager.default.temporaryDirectory
      .appendingPathComponent("expo-ai-kit-\(UUID().uuidString)")
    do {
      try FileManager.default.moveItem(at: location, to: stableTemp)
      didFinish = true
      continuation?.resume(returning: stableTemp)
      continuation = nil
    } catch {
      didFinish = true
      continuation?.resume(throwing: error)
      continuation = nil
    }
  }

  func urlSession(
    _ session: URLSession,
    downloadTask: URLSessionDownloadTask,
    didWriteData bytesWritten: Int64,
    totalBytesWritten: Int64,
    totalBytesExpectedToWrite: Int64
  ) {
    progressHandler(totalBytesWritten, totalBytesExpectedToWrite)
  }

  func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
    guard !didFinish else { return }
    if let error = error {
      continuation?.resume(throwing: error)
    } else if let httpResp = task.response as? HTTPURLResponse, !(200...299).contains(httpResp.statusCode) {
      continuation?.resume(throwing: NSError(
        domain: "ExpoAiKit.Download",
        code: httpResp.statusCode,
        userInfo: [NSLocalizedDescriptionKey: "HTTP \(httpResp.statusCode)"]
      ))
    } else {
      continuation?.resume(throwing: NSError(
        domain: "ExpoAiKit.Download",
        code: -1,
        userInfo: [NSLocalizedDescriptionKey: "Download finished without file"]
      ))
    }
    continuation = nil
  }
}

// MARK: - Errors

enum GemmaError: LocalizedError {
  case modelNotDownloaded(String)
  case modelLoadFailed(String, reason: String)
  case inferenceFailed(String, reason: String)
  case downloadFailed(String, reason: String)
  case downloadCancelled(String)
  case downloadCorrupt(String, expected: String, actual: String)

  // Error string formatted as "<CODE>:<modelId>:<reason>" to match the
  // contract that the JS layer's ModelError parses.
  var errorDescription: String? {
    switch self {
    case .modelNotDownloaded(let id):
      return "MODEL_NOT_DOWNLOADED:\(id):Model file not found on disk"
    case .modelLoadFailed(let id, let reason):
      return "MODEL_LOAD_FAILED:\(id):\(reason)"
    case .inferenceFailed(let id, let reason):
      return "INFERENCE_FAILED:\(id):\(reason)"
    case .downloadFailed(let id, let reason):
      return "DOWNLOAD_FAILED:\(id):\(reason)"
    case .downloadCancelled(let id):
      return "DOWNLOAD_CANCELLED:\(id):Download cancelled"
    case .downloadCorrupt(let id, let expected, let actual):
      return "DOWNLOAD_CORRUPT:\(id):SHA256 mismatch: expected \(expected), got \(actual)"
    }
  }
}
